_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+What is needed is a way to ensure that the mocked (or stubbed) behavior of the test objects is connected to the actual behavior of the objects being modeled. Ideally, these mocked objects would be defined in the project they represent.
+
+Spring Cloud Contract does exactly this. It provides a way of specifying specific behavior for a service -- the "contract" -- and testing for compliance with this contract **within that project itself** while allowing other projects to load these contracts as stubs that can be mocked for integration testing. 
+
+{% img blog/spring-cloud-contract/spring-cloud-contract-diagram.png alt:"Spring Cloud Contract Diagram" width:"700" %}{: .center-image }
+
+A brief aside on "mocks" and "stubs". "Mock" gets used in ways online that is somewhat generic, meaning any fake object used for testing, and this can get confusing when differentiating "mocks" from "stubs". However, specifically, a "mock" is an object that tests for behavior by registering expected method calls before a test run while a "stub" is a testable version of the object with callable methods that return pre-set values. Thus, a mock checks to see if the object being tested makes an expected sequence of calls to the object being mocked, and throws an error if the behavior deviates (that is, makes any unexpected calls), while a stub does not do any testing itself, per se, but instead will return canned responses to pre-determined methods to allow tests to run (of course, it can throw an error if unexpected calls are made as well, so stubs do constrain behavior in that sense as well). 
+
+In this tutorial, you're going to create a simple Spring Boot **producer** that serves hats and a Spring Boot **consumer** that presents hats to the user (really hat specification object instances). The consumer will retrieve the hats from the producer before presenting them. You will use Spring Cloud Contract to write a contract that describes what the producer service should return based on certain resource queries. Spring Cloud Contract will automatically generate integration tests from the producer contract, which will verify that the producer is behaving according to the specification of the contract, and will also generate stubs that can be imported into other projects to assist in integration testing. Next, on the consumer side, you will import these stubs into the consumer's integration tests as a mocked object that can be used in testing the consumer. You will also see how Okta can be used as an OAuth 2.0 and OpenID Connect (OIDC) provider to add security to the Spring Boot project.
```

Comment:
```suggestion
In this tutorial, you're going to create a simple Spring Boot **producer** that serves hats and a Spring Boot **consumer** that presents hats to the user (really hat specification object instances). The consumer will retrieve the hats from the producer before presenting them. You will use Spring Cloud Contract to write a contract that describes what the producer service should return based on certain resource queries. Spring Cloud Contract will automatically generate integration tests from the producer contract, which will verify that the producer is behaving according to the specification of the contract. It will also generate stubs that can be imported into other projects to assist in integration testing. Next, on the consumer side, you will import these stubs into the consumer's integration tests as a mocked object that can be used in testing the consumer. You will also see how Okta can be used as an OAuth 2.0 and OpenID Connect (OIDC) provider to add security to the Spring Boot project.
```
Breaking up a lengthy sentence for ease-of-reading